test/e2e/integreatly_test.go

```diff
@@ -115,9 +116,33 @@ func integreatlyWorkshopTest(t *testing.T, f *framework.Framework, ctx *framewor
 			break
 		}
 	}
-	//These test only that the operators themselves have came up. Further testing should be done to verify the individual parts of each product come up also (What the operator is creating).
-
-	return err
+	// run infinite loop and only break if the products are installed, e.g. installation is completed
+	for {
+		workshopInstallation = &operator.Installation{}
+		err = f.Client.Get(goctx.TODO(), types.NamespacedName{Name: installationName, Namespace: namespace}, workshopInstallation)
+		if err != nil {
+			return err
+		}
+		products := fmt.Sprintf("%#v", workshopInstallation.Status.Stages["products"])
+		if strings.Contains(products, "completed") {
+			t.Log("Products installed. Waiting for solution-explorer installation to complete")
+			for {
+				workshopInstallation = &operator.Installation{}
+				err = f.Client.Get(goctx.TODO(), types.NamespacedName{Name: installationName, Namespace: namespace}, workshopInstallation)
+				if err != nil {
+					return err
+				}
+				webapp := fmt.Sprintf("%#v", workshopInstallation.Status.Stages["solution-explorer"])
+				if strings.Contains(webapp, "completed") {
+					return err
+				} else {
+					time.Sleep(deploymentRetryInterval)
+				}
+			}
+		} else {
+			time.Sleep(deploymentRetryInterval)
+		}
+	}
```

Comment:
I would have seen these tests going more along the lines of:

- For each stage in the type of installation we are testing (managed, workshop, whatever we might add later)
  - For each operator being installed in this stage
    - Wait with timeout for the operator deployment to succeed(waitForProductDeployment)
  - Wait with timeout for the phase status to be complete (.status.stages.\<phasename\>.phase)

So the current "managed" installation type would be more like:
````
func waitForInstallatonStage() {
	//Implemenet this taking inspiration from https://github.com/operator-framework/operator-sdk/blob/master/pkg/test/e2eutil/wait_util.go#L46-L72
	//Should use a timeout and retryInterval to avoid a test just sitting infinitely waiting for some state
}

//Bootstrap Stage
//verify bootstrap stage completes
err = waitForInstallatonStage(t, f, ctx, installationName, "bootstrap")
if err != nil {
	return err
}

//Auth Stage
//verify operators deploy
err = waitForProductDeployment(t, f, ctx, "rhsso", "keycloak-operator")
if err != nil {
	return err
}
//verify authentication stage completes
err = waitForInstallatonStage(t, f, ctx, installationName, "authentication")
if err != nil {
	return err
}

//Products Stage
//verify operators deploy
for product, deploymentName := range products {
	err = waitForProductDeployment(t, f, ctx, product, deploymentName)
}
if err != nil {
	return err
}
//verify authentication stage completes
err = waitForInstallatonStage(t, f, ctx, installationName, "products")

//Solutions Explorer Stage
//verify operators deploy
err = waitForProductDeployment(t, f, ctx, "solution-explorer", "tutorial-web-app-operator")
if err != nil {
	return err
}
//verify authentication stage completes
err = waitForInstallatonStage(t, f, ctx, installationName, "solution-explorer")
if err != nil {
return err
}
````

This way you are testing each stage consistently and it will be far easier to update as the operator evolves. It will also allow for simpler addition of tests for different installation scenarios should we want to do that, which we likely will if/when we introduce different supported installation configurations.

Reply:
wanna have another look? I have updated the code based on your comments

Reply:
Looks good, pretty much how i seen it working. :+1: